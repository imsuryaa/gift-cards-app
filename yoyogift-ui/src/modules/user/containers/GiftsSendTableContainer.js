import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { Redirect } from 'react-router-dom';
import CircularProgress from '@material-ui/core/CircularProgress';
import { Table, Column, InfiniteLoader } from "react-virtualized";
import 'react-virtualized/styles.css'

import { fetchFilteredSentCards } from "../state/actions";

const mapStateToProps = ({
  login: {
      detailsObject,
      loginStatus,
  }, users: { cards }
}) => {
  return {
      user: detailsObject,
      isLoggedIn: loginStatus,
      sentCards: cards
  }
};

export const mapDispatchToProps = dispatch => {
  return bindActionCreators({
    fetchFilteredSentCards
  },
  dispatch
  );
}

export const GiftsSendTableContainer = ({
  user,
  isLoggedIn,
  sentCards,
  fetchFilteredSentCards
}) => {
  const rowLimit = 100;

  const isRowLoaded = ({ index }) => index < sentCards.length;
  const loadMoreRows = () => {
     fetchFilteredSentCards(user.email, rowLimit, page);
     setPage(page + 1);
  }
  const rowGetter = ({ index }) => sentCards[index];
  const [ page, setPage ] = useState(1);
  useEffect(()=> {
    if(user) {
      fetchFilteredSentCards(user.email, rowLimit, page);
      setPage(page + 1); 
    }
  }, []);

  return(
    <>
      {!isLoggedIn && <Redirect to="/" />}
      {(!sentCards || sentCards.length < 0) && <CircularProgress style={{marginLeft: '50%', marginTop: '10%'}} /> } 
      {sentCards.length === 0 && 
        <h2 style={{
          height: '40px',
          background: '#b3d9f7',
          color: 'white',
          textAlign: 'center',
          verticalAlign: 'middle',
          paddingTop: '5px',
          fontSize: '20px',
          fontWeight: '500'
        }}>
        NO DATA
        </h2> }
      {sentCards.length > 0 &&
        <div>
          <InfiniteLoader
            isRowLoaded={isRowLoaded}
            rowCount={sentCards.length+1}
            loadMoreRows={loadMoreRows}
          >
            {({ onRowsRendered, registerChild }) => (
              <Table
                ref={registerChild}
                rowCount={sentCards.length}
                rowGetter={rowGetter}
                headerHeight={50}
                rowHeight={50}
                overscanRowCount={1}
                width={1350}
                height={495}
                onRowsRendered={onRowsRendered}
              >
                <Column dataKey="id" label="ID" width={100} />
                <Column dataKey="cardName" label="CARD NAME" width={200} />
                <Column dataKey="cardPoints" label="POINTS" width={200} />
                <Column dataKey="senderEmail" label="SENT TO" width={300} />
                <Column dataKey="cardIssueDate" label="ISSUE DATE" width={275} />
                <Column dataKey="cardExpiryDate" label="EXPIRY DATE" width={275} />
              </Table>
            )}
          </InfiniteLoader>
        </div>
      }
    </>
  );
};

GiftsSendTableContainer.propTypes = {
  user: PropTypes.object.isRequired,
  isLoggedIn: PropTypes.bool.isRequired,
  sentCards: PropTypes.array.isRequired,
  fetchFilteredSentCards: PropTypes.func.isRequired
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GiftsSendTableContainer);