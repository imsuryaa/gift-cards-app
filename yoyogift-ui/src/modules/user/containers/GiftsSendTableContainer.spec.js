import React from "react";
import { mount } from "enzyme";

import { GiftsSendTableContainer, mapDispatchToProps } from "./GiftsSendTableContainer";

describe("Gifts Sent Table Test Suite", () => {
  it('should show header with information', () => {
    const wrapper = mount(<GiftsSendTableContainer 
    isLoggedIn = {true}
    sentCards={[]}
    fetchFilteredSentCards={jest.fn()}
    />);

    expect(wrapper.find('h2').length).toBe(1);
    expect(wrapper.find('h2').text()).toBe("NO DATA");
  });

  it('should show table with header and data', () => {
    //Mocks
    const data = [
        {
          cardName: 'A',
          cardPoints: 1,
          receiverEmail: 'jalagamsurya.loriini@gmail.com',
          cardIssueDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)",
          cardExpiryDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)"
        },
        {
          cardName: 'B',
          cardPoints: 2,
          receiverEmail: 'jalagamsurya.loriini@gmail.com',
          cardIssueDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)",
          cardExpiryDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)"
        },
        {
          cardName: 'C',
          cardPoints: 3,
          receiverEmail: 'jalagamsurya.loriini@gmail.com',
          cardIssueDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)",
          cardExpiryDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)"
        }
    ];

    const user = {
      email: "jalagamsurya.loriini@gmail.com"
    };

    const wrapper = mount(<GiftsSendTableContainer 
    isLoggedIn = {true}
    sentCards={data}
    user={user}
    fetchFilteredSentCards={jest.fn()}
    />);
    
    expect(wrapper.find('div').length).toBe(32);
  });

  it('should render the expected HTML', () => {
    expect(mount(<GiftsSendTableContainer 
    isLoggedIn = {true}
    sentCards={[]}
    fetchFilteredSentCards={jest.fn()}
    />
    ).html()).toMatchSnapshot();
  });

  it('should render the expected HTML', () => {
    //Mocks
    const data = [
        {
          cardName: 'A',
          cardPoints: 1,
          receiverEmail: 'jalagamsurya.loriini@gmail.com',
          cardIssueDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)",
          cardExpiryDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)"
        },
        {
          cardName: 'B',
          cardPoints: 2,
          receiverEmail: 'jalagamsurya.loriini@gmail.com',
          cardIssueDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)",
          cardExpiryDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)"
        },
        {
          cardName: 'C',
          cardPoints: 3,
          receiverEmail: 'jalagamsurya.loriini@gmail.com',
          cardIssueDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)",
          cardExpiryDate: "Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)"
        }
    ];

    const user = {
      email: "jalagamsurya.loriini@gmail.com"
    };

    expect(mount(<GiftsSendTableContainer 
    isLoggedIn = {true}
    user={user}
    sentCards={data}
    fetchFilteredSentCards={jest.fn()}
    />
    ).html()).toMatchSnapshot();
  });

  it("should fetch sent cards", () => {
    let dispatch = jest.fn();

    let props = mapDispatchToProps(dispatch);
    props.fetchFilteredSentCards("jalagamsurya.loriini@gmail.com", 100, 1);

    expect(dispatch).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledTimes(1);
  });
});
