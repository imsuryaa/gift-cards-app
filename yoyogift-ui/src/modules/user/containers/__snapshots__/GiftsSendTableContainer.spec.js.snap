// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gifts Sent Table Test Suite should render the expected HTML 1`] = `"<h2 style=\\"height: 40px; background: rgb(179, 217, 247); color: white; text-align: center; vertical-align: middle; padding-top: 5px; font-size: 20px; font-weight: 500;\\">NO DATA</h2>"`;

exports[`Gifts Sent Table Test Suite should render the expected HTML 2`] = `"<div><div aria-colcount=\\"6\\" aria-rowcount=\\"3\\" class=\\"ReactVirtualized__Table\\" role=\\"grid\\"><div class=\\"ReactVirtualized__Table__headerRow\\" role=\\"row\\" style=\\"height: 50px; overflow: hidden; padding-right: 0px; width: 1350px;\\"><div class=\\"ReactVirtualized__Table__headerColumn\\" role=\\"columnheader\\" style=\\"flex: 0 1 100px;\\"><span class=\\"ReactVirtualized__Table__headerTruncatedText\\" title=\\"ID\\">ID</span></div><div class=\\"ReactVirtualized__Table__headerColumn\\" role=\\"columnheader\\" style=\\"flex: 0 1 200px;\\"><span class=\\"ReactVirtualized__Table__headerTruncatedText\\" title=\\"CARD NAME\\">CARD NAME</span></div><div class=\\"ReactVirtualized__Table__headerColumn\\" role=\\"columnheader\\" style=\\"flex: 0 1 200px;\\"><span class=\\"ReactVirtualized__Table__headerTruncatedText\\" title=\\"POINTS\\">POINTS</span></div><div class=\\"ReactVirtualized__Table__headerColumn\\" role=\\"columnheader\\" style=\\"flex: 0 1 300px;\\"><span class=\\"ReactVirtualized__Table__headerTruncatedText\\" title=\\"SENT TO\\">SENT TO</span></div><div class=\\"ReactVirtualized__Table__headerColumn\\" role=\\"columnheader\\" style=\\"flex: 0 1 275px;\\"><span class=\\"ReactVirtualized__Table__headerTruncatedText\\" title=\\"ISSUE DATE\\">ISSUE DATE</span></div><div class=\\"ReactVirtualized__Table__headerColumn\\" role=\\"columnheader\\" style=\\"flex: 0 1 275px;\\"><span class=\\"ReactVirtualized__Table__headerTruncatedText\\" title=\\"EXPIRY DATE\\">EXPIRY DATE</span></div></div><div aria-label=\\"grid\\" class=\\"ReactVirtualized__Grid ReactVirtualized__Table__Grid\\" role=\\"rowgroup\\" style=\\"box-sizing: border-box; direction: ltr; height: 445px; position: relative; width: 1350px; will-change: transform; overflow-x: hidden; overflow-y: hidden;\\" tabindex=\\"0\\"><div class=\\"ReactVirtualized__Grid__innerScrollContainer\\" role=\\"rowgroup\\" style=\\"width: auto; height: 150px; max-width: 1350px; max-height: 150px; overflow: hidden; position: relative;\\"><div aria-rowindex=\\"1\\" class=\\"ReactVirtualized__Table__row\\" role=\\"row\\" style=\\"height: 50px; left: 0px; position: absolute; top: 0px; width: 1350px; overflow: hidden; padding-right: 0px;\\"><div aria-colindex=\\"1\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 100px;\\" title=\\"\\"></div><div aria-colindex=\\"2\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 200px;\\" title=\\"A\\">A</div><div aria-colindex=\\"3\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 200px;\\" title=\\"1\\">1</div><div aria-colindex=\\"4\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 300px;\\" title=\\"\\"></div><div aria-colindex=\\"5\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 275px;\\" title=\\"Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)\\">Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)</div><div aria-colindex=\\"6\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 275px;\\" title=\\"Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)\\">Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)</div></div><div aria-rowindex=\\"2\\" class=\\"ReactVirtualized__Table__row\\" role=\\"row\\" style=\\"height: 50px; left: 0px; position: absolute; top: 50px; width: 1350px; overflow: hidden; padding-right: 0px;\\"><div aria-colindex=\\"1\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 100px;\\" title=\\"\\"></div><div aria-colindex=\\"2\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 200px;\\" title=\\"B\\">B</div><div aria-colindex=\\"3\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 200px;\\" title=\\"2\\">2</div><div aria-colindex=\\"4\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 300px;\\" title=\\"\\"></div><div aria-colindex=\\"5\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 275px;\\" title=\\"Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)\\">Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)</div><div aria-colindex=\\"6\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 275px;\\" title=\\"Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)\\">Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)</div></div><div aria-rowindex=\\"3\\" class=\\"ReactVirtualized__Table__row\\" role=\\"row\\" style=\\"height: 50px; left: 0px; position: absolute; top: 100px; width: 1350px; overflow: hidden; padding-right: 0px;\\"><div aria-colindex=\\"1\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 100px;\\" title=\\"\\"></div><div aria-colindex=\\"2\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 200px;\\" title=\\"C\\">C</div><div aria-colindex=\\"3\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 200px;\\" title=\\"3\\">3</div><div aria-colindex=\\"4\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 300px;\\" title=\\"\\"></div><div aria-colindex=\\"5\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 275px;\\" title=\\"Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)\\">Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)</div><div aria-colindex=\\"6\\" class=\\"ReactVirtualized__Table__rowColumn\\" role=\\"gridcell\\" style=\\"overflow: hidden; flex: 0 1 275px;\\" title=\\"Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)\\">Sun May 19 2019 15:43:25 GMT+0530 (India Standard Time)</div></div></div></div></div></div>"`;
